lib/spack/spack/cmd/graph.py

```diff
@@ -41,36 +48,46 @@ def setup_parser(subparser):
     )
 
     subparser.add_argument(
-        "-i",
-        "--installed",
-        action="store_true",
-        help="graph installed specs, or specs in the active env (implies --dot)",
+        "-i", "--installed", action="store_true", help="graph specs from the DB"
     )
 
     arguments.add_common_arguments(subparser, ["deptype", "specs"])
 
 
 def graph(parser, args):
-    if args.installed and args.specs:
-        tty.die("cannot specify specs with --installed")
+    env = ev.active_environment()
+    if args.installed and env:
+        tty.die("cannot use --installed with an active environment")
+
+    if args.installed and args.ascii:
+        tty.die("cannot use --installed with --ascii")
+
+    if env and args.ascii:
+        tty.die("cannot use --ascii with an active environment")
 
     if args.color and not args.dot:
         tty.die("the --color option can be used only with --dot")
 
     if args.installed:
         args.dot = True
-        env = ev.active_environment()
-        if env:
-            specs = env.concrete_roots()
-        else:
+        if not args.specs:
             specs = spack.store.STORE.db.query()
+        else:
+            result = []
+            for item in args.specs:
+                result.extend(spack.store.STORE.db.query(item))
+            specs = list(set(result))
+    elif env:
+        args.dot = True
+        specs = env.concrete_roots()
+        if args.specs:
+            specs = [x for x in specs if any(x.satisfies(req) for req in args.specs)]
```

Comment:
This is still not how it behaves elsewhere. Just do `matching_specs`?

Reply:
Now it's using `env.all_matching_specs`